File: algorithm/LeetCode/Subsets II/Subsets II.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
          ret.clear();
          tmp.clear();
          rawVec = nums;
          sort(rawVec.begin(), rawVec.end());
          dfs(0);
          return ret;
    }

    void dfs(int start)
    {
        ret.push_back(tmp); //首先push一个空集
        for(int i = start; i < rawVec.size(); ++i){
            if(i != start && rawVec[i] == rawVec[i - 1]) continue;  //这里的i != start；应该注意
            tmp.push_back(rawVec[i]);
            dfs(i + 1);
            tmp.pop_back();
        }
    }
private:
    vector<vector<int>> ret;
    vector<int> tmp;
    vector<int> rawVec;

};

File: algorithm/LeetCode/Subsets II/Subsets II.java
public class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        List<Integer> tmp = new ArrayList<Integer>();    
        dfs(ret, tmp, nums, 0, nums.length);
        return ret;
    }
    
    public void dfs(List<List<Integer>> ret, List<Integer> tmp, int [] nums, int start, int limit){
        if(start > limit)
            return;
        else if(start == limit){
            ret.add(new ArrayList<Integer>(tmp));
            return;
        }else{
            ret.add(new ArrayList<Integer>(tmp));
            for(int i = start; i < limit; ++i){
                tmp.add(nums[i]);
                dfs(ret, tmp, nums, i+1, limit);
                tmp.remove((Integer)nums[i]);
                while(i+1 < limit && nums[i+1] == nums[i]) //跳过所有的重复的数字
                    i++; //这里加完了之后while循环又会再加一次，正好跳过所有的重复的值
            }
        }
    }
}